From 6ff9f6a8dbbb5514ea8eff49779d5568774cf12f Mon Sep 17 00:00:00 2001
From: Hariprasad Kelam <hkelam@marvell.com>
Date: Fri, 26 Mar 2021 10:11:50 +0530
Subject: [PATCH 04/23] octeontx2-pf: Read correct number of VFs

commit 013a008e29ce56819f295a778b2d80ee8f9a505d from
git@git.assembla.com:cavium/WindRiver.linux.git

In function sriov_disable , while freeing resources number of
vfs is read from pci_num_vf(). But this api returning value as 0
due to pci_disable_sriov internally clears number of vfs.
Fix this by calling pci_disable_sriov after reading pci_num_vf().

Change-Id: Ia453dd607b3acdc5664c6a704fd621d8239102eb
Fixes: 37a6d8b4ab60 ("octeontx2-pf: Notify VFs about link changes")
Signed-off-by: Hariprasad Kelam <hkelam@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/48942
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/52276
Reviewed-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
index 887bcc6ac49b..84c3c659fc44 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
@@ -2653,12 +2653,12 @@ static int otx2_sriov_disable(struct pci_dev *pdev)
 	if (!numvfs)
 		return 0;
 
-	pci_disable_sriov(pdev);
-
 	for (i = 0; i < pci_num_vf(pdev); i++) {
 		cancel_delayed_work_sync(&pf->vf_configs[i].link_event_work);
 		cancel_delayed_work_sync(&pf->vf_configs[i].ptp_info_work);
 	}
+
+	pci_disable_sriov(pdev);
 	kfree(pf->vf_configs);
 
 	otx2_disable_flr_me_intr(pf);
-- 
2.31.1

